### PR TITLE
[HybridParallel]Add gpt example using dygraph hybrid parallel

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

Others
### PR changes

Others
### Description

[HybridParallel]Add gpt example using dygraph hybrid parallel
添加动态图混合并行的gpt3代码示例。精度和单卡精度对齐。

globalsize=256，8机64卡性能数据：

模型配置(layer, hidden_size,head) | 精度 | 策略配置 | Megatron | paddle动态图 | paddle静态图 (Fake 数据）
-- | -- | -- | -- | -- | --
7B_16_6144_128 | fp32 | dp1_pp8_mp8 | 11647 | 11485(-1.4%) | 11593
  | fp16 | dp1_pp8_mp8 | 47801 | 47300(-1%) | 40644
14B_32_6144_128 | fp32 | dp1_pp8_mp8 | OOM | OOM | 6634
  |   | dp1_pp8_mp8_recompute | 5156 | 5865(+13.7%) | 5787
14B_32_6144_128 | fp16 | dp1_pp8_mp8 | 27911 | 27429(-1.8%) | 23261
  |   | dp1_pp8_mp8_recompute | 21751 | 22592(+3.8%) | 17864

### CODE REVIEW

File: examples/language_model/gpt-3/dygraph/args.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    # parser.add_argument(
+    #     "--global_batch_size",
+    #     default=None,
+    #     type=int,
+    #     help="Global batch size for all training process. None for not check the size is valid. If we only use data parallelism, it should be device_num * micro_batch_size."
+    # )
+
+    parser.add_argument(
+        "--local_batch_size",
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/args.py:96]: Why `local_batch_size`?

File: examples/language_model/gpt-3/dygraph/dataset.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+        assert doc_idx.dtype == np.int32
+        sample_idx = _build_sample_idx(sizes, doc_idx, seq_length, num_epochs,
+                                       tokens_per_epoch)
+        if separate_last_epoch:
+            num_samples_ = num_samples_from_epochs_minus_one
+        else:
+            num_samples_ = sample_idx.shape[0] - 1
+        shuffle_idx = _build_shuffle_idx(num_samples_, sample_idx.shape[0] - 1,
+                                         np_rng)
+        if paddle.distributed.get_rank() % 8 == 0:
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/dataset.py]: `8` for hard code here.  you can pass local_rank as  https://github.com/PaddlePaddle/PaddleNLP/pull/930

File: examples/language_model/gpt-3/dygraph/dataset.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            name="gpt_" + name,
+            max_seq_len=max_seq_len,
+            num_samples=num_samples,
+            documents=np.arange(splits[index], splits[index + 1]),
+            sample_ids=sample_ids,
+            sample_lens=sample_lens,
+            eos_id=eos_id,
+            seed=args.seed)
+
+        batch_sampler = paddle.io.DistributedBatchSampler(
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/dataset.py]: If we don't use shuffle, here can use
```
from paddlenlp.utils.batch_sampler import DistributedBatchSampler
```
As https://github.com/PaddlePaddle/PaddleNLP/pull/930/files

`paddlenlp.utils.batch_sampler` could save more memory.

File: examples/language_model/gpt-3/dygraph/dataset.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    # Filename of the index mappings.
+    _filename = data_prefix
+    _filename += '_{}_indexmap'.format(name)
+    _filename += '_{}ns'.format(num_samples)
+    _filename += '_{}sl'.format(seq_length)
+    doc_idx_filename = _filename + '_doc_idx.npy'
+    sample_idx_filename = _filename + '_sample_idx.npy'
+    shuffle_idx_filename = _filename + '_shuffle_idx.npy'
+
+    # support multi-machines
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/dataset.py]: How about save the seed and recover seed later? 
多机我可能不太好验证，要不你先改一下这个seed的方案试试？
怕后面把你的东西改崩了

File: examples/language_model/gpt-3/dygraph/run.sh [hunk 5]
```diff
@@ -0,0 +1,57 @@
+#wget https://paddlenlp.bj.bcebos.com/models/transformers/gpt/train.data.json_ids.npz
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run.sh]: delete

File: examples/language_model/gpt-3/dygraph/run.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+#mkdir data
+#mv train.data.json_ids.npz data
+
+export DATA_DIR=./data
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run.sh]: delete

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+        "pp_degree": args.pp_degree
+    }
+
+    assert args.local_batch_size % args.micro_batch_size == 0
+    strategy.pipeline_configs = {
+        "accumulate_steps": args.local_batch_size // args.micro_batch_size,
+        "micro_batch_size": args.micro_batch_size
+    }
+
+    fleet.init(is_collective=True, strategy=strategy)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:103]: 这一大堆的种子设置，可否用封装成一个辅助工具函数，方便复用？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
+    model_class, tokenizer_class = MODEL_CLASSES[args.model_type]
+    tokenizer = tokenizer_class.from_pretrained(args.model_name_or_path)
+
+    # Define log writer
+    log_writer_path = os.path.join(
+        args.output_dir, "train_log",
+        "{}_globalbsz_{}_amp_{}_recompute_{}_card_{}".format(
+            args.model_name_or_path, default_global_batch_size, args.use_amp,
+            False, worker_index).lower())
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 前面已经设置`worker_index = dp_rank` 对于属于同一dp的mp，这里的日志会不会写入两次？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+                        output_dir = os.path.join(args.output_dir,
+                                                  "model_%d" % global_step)
+                        if not os.path.exists(output_dir):
+                            os.makedirs(output_dir)
+
+                        if args.dp_degree * args.mp_degree * args.pp_degree != 1:
+                            model_to_save = model._layers
+                        else:
+                            model_to_save = model
+                        logger.info("Save model to %s" % output_dir)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: MP 情形下的 save。是否有问题？mp的每个部分应该需要save成不同名字？

**Review by ZHUI — CHANGES_REQUESTED** [review]
很棒的工作👍🏻

File: examples/language_model/gpt-3/dygraph/args.py (hunk 1 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/args.py:96]: global_batch_size 应该是等于local_batch_size *  dp_degree
mircro_batch 是指pp训练为了流水线的性能，将local_batch_size切分位多个小batch。local_batch_size = mircro_batch * accumulate_step
global_batch_size = local_batch_size * dp_degree

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 8 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 是的。我修复一下。

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 7 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:103]: ok, done

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 9 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 是的，需要修复。保存到不同的名字/文件夹。

File: examples/language_model/gpt-3/dygraph/run.sh (hunk 6 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run.sh]: done

File: examples/language_model/gpt-3/dygraph/run.sh (hunk 5 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run.sh]: done

File: examples/language_model/gpt-3/dygraph/dataset.py (hunk 3 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/dataset.py]: done， thx

File: examples/language_model/gpt-3/dygraph/dataset.py (hunk 2 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/dataset.py]: ok， fix it

File: examples/language_model/gpt-3/dygraph/args.py (hunk 1 above)
Comment (ZHUI) [reply examples/language_model/gpt-3/dygraph/args.py:96]: global_batch_size = micro_batch_size * accumulate_step * dp_degree (* sharding_degree)
其实这里设置成 `global_batch_size` 也可以算出，`accumulate_step` 自动去做 accumulate。

设置 global_batch_size 的一个好处是，方便恢复继续训练。只要`global_batch_size`一致，无论单机、多机，相同step保存的状态是一样的。

File: examples/language_model/gpt-3/dygraph/args.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        "--enable_addto",
+        type=str2bool,
+        nargs='?',
+        const=True,
+        help="Whether to enable the addto strategy for gradient accumulation or not. This is only used for AMP training."
+    )
+    parser.add_argument(
+        "--scale_loss",
+        type=float,
+        default=128,
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/args.py]: 换成实际使用的默认值吧

File: examples/language_model/gpt-3/dygraph/run.sh [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    --local_batch_size 16\
+    --dp_degree 2\
+    --mp_degree 4\
+    --pp_degree 1\
+    --use_amp True\
+    --scale_loss 32768
+
+# just for performance
+
+#nsys profile --stats=true -t cuda python -m paddle.distributed.launch --log_dir dp2_pp1_mp4 --gpus "0,1,2,3,4,5,6,7" run_pretrain.py \
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run.sh]: delete

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+                data_world_size=args.dp_degree,
+                data_world_rank=dp_rank,
+                eos_id=tokenizer.eos_token_id)
+            # Bug fix, if not call valid_data_loader, the enumerate will call valid_data_loader
+            # many times. and start a new random dataloader.
+            valid_data_loader = valid_data_loader()
+            test_data_loader = test_data_loader()
+
+            for step, batch in enumerate(train_data_loader()):
+                global_step += 1
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:239]: 如果有 accumulate_step 的话，这里的 global_step 应该经过accumulate_step 再加一。
对应的 lr_scheduler.step() 应该是 每次 global_step 变化的时候，再更新。

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+
+                    if args.use_amp:
+                        scaler.scale(loss).backward()
+                        scaler.minimize(optimizer, loss)
+                    else:
+                        loss.backward()
+                        optimizer.step()
+
+                    if lr_scheduler is not None:
+                        lr_scheduler.step()
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:266]: 如上，lr_scheduler 这里应该是，跟随 global_step 去调整，不随着 accumulate_step 影响，

File: examples/language_model/gpt-3/dygraph/run.sh [hunk 14]
```diff
@@ -0,0 +1,51 @@
+export PYTHONPATH=$PYTHONPATH:../../../../
+
+rm -rf dp2_pp2_mp2
+export NCCL_DEBUG=INFO
+#export NCCL_DEBUG_SUBSYS=ALL
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run.sh]: delete

File: examples/language_model/gpt-3/dygraph/run.sh [hunk 15]
```diff
@@ -0,0 +1,51 @@
+export PYTHONPATH=$PYTHONPATH:../../../../
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run.sh]: delete

**Review by ZHUI — CHANGES_REQUESTED** [review]

File: examples/language_model/gpt-3/dygraph/args.py (hunk 1 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/args.py:96]: 嗯嗯，是的。已修改。添加了global_batch_size,并对这些关系做了判断，可以指定global_batch_size而不需要指定local_batch_size了。

File: examples/language_model/gpt-3/dygraph/args.py (hunk 10 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/args.py]: ok, done

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 12 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:239]: 额，现在的逻辑是每次读local_batch_size,然后框架再去切分microbatchsize，目前这个是和静态图一致

File: examples/language_model/gpt-3/dygraph/run.sh (hunk 15 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run.sh]: done

File: examples/language_model/gpt-3/dygraph/run.sh (hunk 14 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run.sh]: done

**Review by ZHUI — APPROVED** [review]
LGTM

**Review by ZeyuChen — APPROVED** [review]
LGTM!
